rclpy/src/rclpy/node.cpp

```diff
@@ -62,4 +97,75 @@ get_node_namespace(py::capsule pynode)
 
   return node_namespace;
 }
+
+py::list
+get_node_names_impl(py::capsule pynode, bool get_enclaves)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcutils_string_array_t node_names = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t node_namespaces = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t enclaves = rcutils_get_zero_initialized_string_array();
+
+  rcl_ret_t ret = RCL_RET_OK;
+  if (get_enclaves) {
+    ret = rcl_get_node_names_with_enclaves(
+      node, allocator, &node_names, &node_namespaces, &enclaves);
+  } else {
+    ret = rcl_get_node_names(
+      node, allocator, &node_names, &node_namespaces);
+  }
+  if (RCL_RET_OK != ret) {
+    throw RCLError("Failed to get node names");
+  }
+
+  py::list pynode_names_and_namespaces(node_names.size);
+
+  size_t idx;
+  for (idx = 0; idx < node_names.size; ++idx) {
+    if (get_enclaves) {
+      pynode_names_and_namespaces[idx] = py::make_tuple(
+        py::str(node_names.data[idx]),
+        py::str(node_namespaces.data[idx]),
+        py::str(enclaves.data[idx]));
+    } else {
+      pynode_names_and_namespaces[idx] = py::make_tuple(
+        py::str(node_names.data[idx]),
+        py::str(node_namespaces.data[idx]));
+    }
+  }
+
+  rcutils_ret_t rcutils_ret;
+  rcutils_ret = rcutils_string_array_fini(&node_names);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    throw RCLError("Failed to destroy node names");
+  }
+  rcutils_ret = rcutils_string_array_fini(&node_namespaces);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    throw RCLError("Failed to destroy node_namespaces");
+  }
+  rcutils_ret = rcutils_string_array_fini(&enclaves);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    throw RCLError("Failed to destroy enclaves string array");
+  }
```

Comment:
@gbalke consider putting all these "fini" calls in an `RCPPUTILS_SCOPE_EXIT` block *before* populating the list, in case something throws while doing so.

Reply:
Along that same line, make sure to finalize everything before throwing.

Reply:
If I put it in before populating the list then won't the data be destroyed? Or am I missunderstanding?

Reply:
> Along that same line, make sure to finalize everything before throwing from this block.

Good call.

Reply:
@hidmic I have updated to account for these issues.